File: Java/875.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int right = 0;
        for (int i : piles) {
            if (right < i) {
                right = i;
            }
        }
        int left = 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            int currentPace = currentPace(mid, piles);
            if (currentPace <= h) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }

    private int currentPace(int k, int[] piles) {
        int sum = 0;
        for (int i : piles) {
            sum += Math.ceil((double) i / k);
        }
        return sum;
    }
}

File: Python/875.py
import math
from typing import List


class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:

        def hours_required(k):
            total = 0
            for pile in piles:
                total += math.ceil(float(pile) / k)
            return total

        left, right = 1, max(piles)
        result = right
        while left <= right:
            mid = left + (right - left) // 2
            require = hours_required(mid)
            if require <= h:
                result = mid
                right = mid - 1
            else:
                left = mid + 1
        return result
